File: 2024/day01/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

func readInput() ([]int, []int, error) {
	file, err := os.Open("../input.txt")
	if err != nil {
		return nil, nil, err
	}
	defer file.Close()

	var leftList, rightList []int
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Fields(line)
		if len(parts) != 2 {
			continue
		}

		left, err := strconv.Atoi(parts[0])
		if err != nil {
			return nil, nil, err
		}

		right, err := strconv.Atoi(parts[1])
		if err != nil {
			return nil, nil, err
		}

		leftList = append(leftList, left)
		rightList = append(rightList, right)
	}

	return leftList, rightList, scanner.Err()
}

func part1(leftList, rightList []int) int {
	// Create copies to avoid modifying originals
	left := make([]int, len(leftList))
	right := make([]int, len(rightList))
	copy(left, leftList)
	copy(right, rightList)

	// Sort both lists
	sort.Ints(left)
	sort.Ints(right)

	// Calculate total distance
	totalDistance := 0
	for i := 0; i < len(left); i++ {
		totalDistance += int(math.Abs(float64(left[i] - right[i])))
	}

	return totalDistance
}

func part2(leftList, rightList []int) int {
	// Count occurrences in right list
	rightCounts := make(map[int]int)
	for _, num := range rightList {
		rightCounts[num]++
	}

	// Calculate similarity score
	similarityScore := 0
	for _, num := range leftList {
		similarityScore += num * rightCounts[num]
	}

	return similarityScore
}

func main() {
	leftList, rightList, err := readInput()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Part 1: %d\n", part1(leftList, rightList))
	fmt.Printf("Part 2: %d\n", part2(leftList, rightList))
}

File: 2024/day01/c/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINES 1000

// Comparison function for qsort
int compare_ints(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

int main() {
    FILE *file = fopen("../input.txt", "r");
    if (!file) {
        perror("Failed to open input.txt");
        return 1;
    }

    int left[MAX_LINES], right[MAX_LINES];
    int count = 0;

    // Read input
    while (fscanf(file, "%d %d", &left[count], &right[count]) == 2) {
        count++;
        if (count >= MAX_LINES) break;
    }
    fclose(file);

    // Part 1: Sort both lists and calculate total distance
    int left_sorted[MAX_LINES], right_sorted[MAX_LINES];
    memcpy(left_sorted, left, count * sizeof(int));
    memcpy(right_sorted, right, count * sizeof(int));

    qsort(left_sorted, count, sizeof(int), compare_ints);
    qsort(right_sorted, count, sizeof(int), compare_ints);

    long long part1_sum = 0;
    for (int i = 0; i < count; i++) {
        int diff = left_sorted[i] - right_sorted[i];
        part1_sum += diff > 0 ? diff : -diff;
    }

    // Part 2: Calculate similarity score
    // For each number in left list, count occurrences in right list
    long long part2_sum = 0;
    for (int i = 0; i < count; i++) {
        int num = left[i];
        int occurrences = 0;
        for (int j = 0; j < count; j++) {
            if (right[j] == num) {
                occurrences++;
            }
        }
        part2_sum += (long long)num * occurrences;
    }

    printf("Part 1: %lld\n", part1_sum);
    printf("Part 2: %lld\n", part2_sum);

    return 0;
}
